posts/2021/07/moving-scipy-to-meson.md

```diff
@@ -0,0 +1,489 @@
+[Meson](https://mesonbuild.com/) on Linux, and the full test suite passes!**_
+
+This is a pretty exciting milestone, and good news for SciPy maintainers and
+contributors - they can look forward to much faster builds and a more
+pleasant development experience. So how fast is it? Currently the build takes
+about 1min 50s on my 3 year old 12-core Intel CPU (i9-7920X @ 2.90GHz):
+
+![Profiling result of a parallel build of SciPy with Meson](/images/2021/07/ninjabuild_tracing_12jobs.png)
+
+*Profiling result of a parallel build (12 jobs) of SciPy with Meson. Visualization created with [ninjatracing](https://github.com/nico/ninjatracing/blob/master/ninjatracing) and [Perfetto](https://ui.perfetto.dev).*
+
+As you can see from the tracing results, building a single C++ file
+(`bsr.cxx`, one of SciPy's sparse matrix formats) takes over 90 seconds. So
+short of major surgery on that C++ code or buying a newer CPU, the 1min 50sec
+is close to optimal.
+
+<!-- TEASER_END -->
+
+## Why move to Meson, and why now?
+
+In Python 3.10 `distutils` will be deprecated and in Python 3.12 it will be
+removed (see [PEP 632](https://www.python.org/dev/peps/pep-0632/)). SciPy
+uses `numpy.distutils` and `setuptools` as its build system, so the removal
+of `distutils` from the Python standard library would have a major impact on
+SciPy - and even more on `numpy.distutils`, which directly extends
+`distutils`. That PEP was written almost a year ago, and at the time my first
+instinct was to wait until `setuptools` integrated `distutils` (which the
+`setuptools` maintainers still plan to do, with a cleaned up API) and then
+update `numpy.disutils` for that change. That would also require moving parts
+of `numpy.distutils` into `setuptools`, like Fortran support (see
+[setuptools/issues/2372](https://github.com/pypa/setuptools/issues/2372) for
+more details). It has become clear though that this will be a really slow and
+painful process - after almost a year, the vendored `distutils` still hasn't
+been re-enabled inside `setuptools`.
+
+The other driver was that SciPy development has become more painful over
+time. SciPy contains C, C++, Fortran and Cython code, exports C and Cython
+APIs, has complex dependencies (see diagram below) and does a lot of code
+generation. The growing amount of C++ and Cython code has increased build
+times. CI systems frequently time out because build plus test suite run takes
+close to 60 minutes (the limit on several free-for-open-source CI offerings).
+Working on compiled code is cumbersome and slow, to the point where a couple
+of other maintainers have told me it's a blocker for them to work on new
+features. Finally, debugging build issues is quite difficult - `distutils`
+doesn't really have a design, so every other extension just monkey patches
+things as needed. Earlier in the year I spent half a day hunting a build
+problem through four (!) different projects that all modified how extensions
+are built.
```

Comment:
Nice to have: is there a  link to this work?

Reply:
I can't find it so quickly - it was somewhere in the Pythran integration PRs. I'm not sure there's much to learn from that, other than "this is a giant mess".